viewing_party/party.py

```diff
@@ -1,23 +1,263 @@
+    all_friends=user_data["friends"]
+
+    for f in all_friends:
+        friend_unique_movies = find_unique_user_movies(f)
+        list_of_titles.extend(friend_unique_movies)
+        list_all_movies.extend(find_user_movies(f))
+
+    set_of_unique_titles_friends=set(list_of_titles)
+    set_user_watched = find_unique_user_movies(user_data)
+
+    movies_only_friends_watched=set_of_unique_titles_friends.difference(set_user_watched)
+
+    output=[]
+    for m in list_all_movies:
+        if m["title"] in movies_only_friends_watched:
+            output.append(m)
+            movies_only_friends_watched.remove(m["title"])
+    print (output)  
+    return output      
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+#Determine a list of recommended movies. A movie should be added to this list if and only if:
+# The user has not watched it
+# At least one of the user's friends has watched
+# The `"host"` of the movie is a service that is in the user's `"subscriptions"`
+def get_available_recs(user_data):
+    recom_movies=[]
+    unique_movies=get_friends_unique_watched(user_data)
+    for movie in unique_movies:
+        if movie["host"] in user_data["subscriptions"]:
+            recom_movies.append(movie)
+    return recom_movies        
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+#Consider the user's most frequently watched genre. Then, determine a list of recommended movies. A movie should be added to this list if and only if:
+# - The user has not watched it
+# - At least one of the user's friends has watched
+# - The `"genre"` of the movie is the same as the user's most frequent genre
+def get_new_rec_by_genre(user_data):
+    recommended_movies=[]
+    dict_of_freq_genre={}
+
+    watched_by_user=user_data["watched"]
+
+    for movie in watched_by_user:
+        if movie["genre"] in dict_of_freq_genre:
+            dict_of_freq_genre[movie["genre"]]+=1
+        else:
+            dict_of_freq_genre[movie["genre"]]=1
+    
+    print(dict_of_freq_genre)
+    most_freq_movie=0
+    most_freq_genre_name = ""
+
+    for genre in dict_of_freq_genre:
+        if dict_of_freq_genre[genre]>most_freq_movie:
+            most_freq_movie=dict_of_freq_genre[genre]
+            most_freq_genre_name = genre
+
+    unique_movies=get_friends_unique_watched(user_data)
+
+    for movie in unique_movies:
+        if movie["genre"]==most_freq_genre_name:
+            recommended_movies.append(movie)
+    return recommended_movies        
+
+def get_movies_user_fav_friends_not_watched(user_data):
+    set_friends_watched=set()
+    list_friend=user_data["friends"]
+    for friend in list_friend:
+        moview_watched_friend=friend["watched"]
+        for movie in moview_watched_friend:
+            title=movie["title"]
+            set_friends_watched.add(title)

```

Comment:
This section is the same as lines 96-101.  Identical code blocks are great candidates for helper functions and can help improve readability by DRY-ing (Don't Repeat Yourself) your code.